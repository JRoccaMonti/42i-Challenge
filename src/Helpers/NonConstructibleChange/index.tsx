const NonConstructibleChange = (input: number[]): number=> {
    const inputShort = input.sort((a, b) => a - b);
    let change = 0;
  
    for (const coin of inputShort) {
      if (coin > change + 1) return change + 1;
      change += coin;
    }
  
    return change + 1;
};
 
export default  NonConstructibleChange;