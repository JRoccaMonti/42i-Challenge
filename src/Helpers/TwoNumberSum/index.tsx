export const twoNumberSum = (input: number[], target: number): number[] => {
  const seen : { [key: number]: number } = {}
  for (let i = 0; i < input.length; i++) {
    const complement = target - input[i];
    if (Object.prototype.hasOwnProperty.call(seen , complement)) {
      return [complement,input[i]];
    }
    seen [input[i]]=i;
  }
  
  return [];
};


