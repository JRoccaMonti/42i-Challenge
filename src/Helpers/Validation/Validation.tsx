export const validateArrayNumbers = (
    numbers: string, 
    validateDuplicates: boolean = false,
    AllowNegatives: boolean = true,
  ): string[] => {
    const errors: string[] = [];
    
    if (numbers.startsWith(',') || numbers.endsWith(',')) {
      errors.push('La lista de números no debe comenzar ni terminar con una coma.');
    }
    if (AllowNegatives) {
      if (/[^0-9-],|,[^0-9-]/.test(numbers)) {
        errors.push('Las comas deben estar separadas por números enteros.');
      }
    }else{
      if (/[^0-9],|,[^0-9]/.test(numbers)) {
        errors.push('Las comas deben estar separadas por números enteros positivos.');
      }
    }
  
    const numArray = numbers.split(',')
      .map(num => num.trim())
      .filter(num => num !== '');
  
    numArray.forEach(num => {
      const parsedNum = parseFloat(num);
      if (isNaN(parsedNum)) {
        errors.push(`El valor "${num}" no es un número válido.`);
      }
    });
  
    if (validateDuplicates) {
      const numSet = new Set(numArray);
      if (numSet.size !== numArray.length) {
        errors.push('La lista de números no debe contener duplicados.');
      }
    }
  
    return errors;
};  
  
export const validateTarget = (target: number | ''): string[] => {
    const errors: string[] = [];
    if (target === '' || !Number.isInteger(target)) {
        errors.push('Por favor, ingresa un objetivo válido (número entero).');
    }
   return errors;
};
  
export const validateNumArray = (numArray: number[]): string[] => {
    const errors: string[] = [];
    if (numArray.length === 0) {
        errors.push('Por favor, ingresa al menos un número entero.');
    }
    return errors;
};
  