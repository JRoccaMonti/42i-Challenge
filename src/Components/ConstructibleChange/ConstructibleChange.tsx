import React, { useState } from 'react';
import NonConstructibleChange from '../../Helpers/NonConstructibleChange';
import { validateArrayNumbers, validateNumArray } from '../../Helpers/Validation/Validation';

const ConstructibleChange: React.FC = () => {
  const [numbers, setNumbers] = useState<string>('');
  const [result, setResult] = useState<number>();
  const [errors, setErrors] = useState<string[]>([]);
  
  const validateInputs = () => {
    const newErrors: string[] = [];
    newErrors.push(...validateArrayNumbers(numbers,false,false));
    const numArray = numbers.split(',')
      .map(num => parseFloat(num.trim()))
      .filter(num => !isNaN(num));
    newErrors.push(...validateNumArray(numArray));
    setErrors(newErrors);
    return newErrors.length === 0;
  };
  const handleCalculate = () => {
    if (!validateInputs()) return;
    const numArray = numbers.split(',').map(num => parseFloat(num.trim())).filter(num => !isNaN(num));
    const sumResult = NonConstructibleChange(numArray);
    setResult(sumResult);
  };

  return (
    <div>
      <div>
        <label>
          NÃºmeros (separados por comas):
          <input
            type="text"
            value={numbers}
            onChange={(e) => setNumbers(e.target.value)}
            placeholder="Ej: 1, 2, 3, 4"
          />
        </label>
      </div>
      <button onClick={handleCalculate}>Calcular</button>
      {errors.length > 0 && (
        <div style={{ color: 'red' }}>
          {errors.map((error, index) => (
            <p key={index}>{error}</p>
          ))}
        </div>
      )}
      <div>
        <p>Cambio minimo : {result}</p>
      </div>
    </div>
  );
};

export default ConstructibleChange;