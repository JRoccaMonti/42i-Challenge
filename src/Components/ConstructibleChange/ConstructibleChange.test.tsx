import { render, fireEvent, screen  } from '@testing-library/react';
import '@testing-library/jest-dom'
import {ConstructibleChange} from "./ConstructibleChange";
import {nonConstructibleChange} from '../../Helpers/nonConstructibleChange';
import { validateArrayNumbers, validateNumArray } from '../../Helpers/Validation/Validation';

jest.mock('../../Helpers/nonConstructibleChange');
jest.mock('../../Helpers/Validation/Validation');

describe('ConstructibleChange Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('Debería coincidir con la snapshot', () => {
    const { asFragment } = render(<ConstructibleChange />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Debería mostrar errores de validación para entradas no válidas', () => {
    (validateArrayNumbers as jest.Mock).mockReturnValue(['Error en la validación de array']);
    (validateNumArray as jest.Mock).mockReturnValue([]);

    render(<ConstructibleChange />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: 'abc, def' } });
    fireEvent.click(screen.getByText('Calcular'));

    expect(screen.getByText('Error en la validación de array')).toBeInTheDocument();
  });

  it('No debería calcular el cambio mínimo si hay errores de validación', () => {
    (validateArrayNumbers as jest.Mock).mockReturnValue(['Error en la validación de array']);
    (validateNumArray as jest.Mock).mockReturnValue([]);

    render(<ConstructibleChange />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: 'abc, def' } });
    fireEvent.click(screen.getByText('Calcular'));

    expect(nonConstructibleChange).not.toHaveBeenCalled();
    expect(screen.queryByText('Cambio minimo :'));
  });

  it('Debería calcular y mostrar el cambio mínimo correctamente', () => {
    (validateArrayNumbers as jest.Mock).mockReturnValue([]);
    (validateNumArray as jest.Mock).mockReturnValue([]);
    (nonConstructibleChange as jest.Mock).mockReturnValue(7);

    render(<ConstructibleChange />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: '1, 2, 5' } });
    fireEvent.click(screen.getByText('Calcular'));

    expect(nonConstructibleChange).toHaveBeenCalledWith([1, 2, 5]);
    expect(screen.getByText('Cambio minimo : 7')).toBeInTheDocument();
  });
  it('Debería actualizar el estado de números al ingresar texto', () => {
    render(<ConstructibleChange />);

    fireEvent.change(screen.getByPlaceholderText('Ej: 1, 2, 3, 4'), { target: { value: '1, 2, 3' } });

    expect(screen.getByPlaceholderText('Ej: 1, 2, 3, 4')).toHaveValue('1, 2, 3');
  });
});
