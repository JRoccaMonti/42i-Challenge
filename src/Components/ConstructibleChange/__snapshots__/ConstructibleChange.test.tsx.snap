// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConstructibleChange Component Debería coincidir con la snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <label>
        Números (separados por comas):
        <input
          placeholder="Ej: 1, 2, 3, 4"
          type="text"
          value=""
        />
      </label>
    </div>
    <button>
      Calcular
    </button>
    <div>
      <p>
        Cambio minimo : 
      </p>
    </div>
  </div>
</DocumentFragment>
`;
